compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -785,23 +785,25 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // Do we have as many extra provided arguments as the tuple's length?
             // If so, we might have just forgotten to wrap some args in a tuple.
             if let Some(ty::Tuple(tys)) =
-                formal_and_expected_inputs.get(mismatch_idx.into()).map(|tys| tys.1.kind())
+                formal_and_expected_inputs.get(mismatch_idx.to_expected_idx()).map(|tys| tys.1.kind())
                 // If the tuple is unit, we're not actually wrapping any arguments.
                 && !tys.is_empty()
                 && provided_arg_tys.len() == formal_and_expected_inputs.len() - 1 + tys.len()
             {
                 // Wrap up the N provided arguments starting at this position in a tuple.
                 let provided_as_tuple = Ty::new_tup_from_iter(
                     tcx,
-                    provided_arg_tys.iter().map(|(ty, _)| *ty).skip(mismatch_idx).take(tys.len()),
+                    provided_arg_tys[mismatch_idx..mismatch_idx.plus(tys.len())]
+                        .iter()
```

Comment:
maybe a mix that avoids the math?

```suggestion
                    provided_arg_tys[mismatch_idx..]
                        .iter()
                        .take(tys.len())
```